components/sensor/ultrasonic/ultrasonic.go

```diff
@@ -109,42 +121,122 @@ type Sensor struct {
 	triggerPin    board.GPIOPin
 	intChan       chan bool
 	timeoutMs     uint
+	readingChan   chan bool
+	distanceChan  chan float64
+	errChan       chan error
+	cancelCtx     context.Context
+	cancelFunc    func()
+	mu            sync.Mutex
+	// toggle this property to be able to lock/unlock
+	// the mutex in sendDistance
+	isReading               bool
+	activeBackgroundWorkers sync.WaitGroup
 	generic.Unimplemented
 }
 
-// Readings returns the calculated distance.
-func (s *Sensor) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+func (s *Sensor) namedError(err error) error {
+	return errors.Wrapf(
+		err, "Error in ultrasonic sensor with name %s: ", s.Name,
+	)
+}
+
+func (s *Sensor) startUpdateLoop(ctx context.Context) {
+	s.activeBackgroundWorkers.Add(1)
+	rdkutils.ManagedGo(
+		func() {
+			defer s.echoInterrupt.RemoveCallback(s.intChan)
```

Comment:
nit: if you do the remove here I think you should register the callback here as well